fixed some issues with kubectl set resources

JacobTanenbaum opened the pull request:
when using kubectl set resources it resets all resource fields that are not being set.

for example
  $ kubectl set resources deployments nginx --limits=cpu=100m

followed by

  $ kubectl set resources deployments nginx --limits=memory=256Mi

would result in the nginx deployment only limiting memory at 256Mi with the previous
limit placed on the cpu being wiped out. This behavior is corrected so that each invocation
only modifies fields set in that command and changed the testing so that the desired behavior
is checked.

Also a typo:

 you must specify an update to requests or limits or  (in the form of --requests/--limits)

corrected to

 you must specify an update to requests or limits (in the form of --requests/--limits)

 Implemented both the dry run and local flags.

Added test cases to show that both flags are operating as intended.
Removed the print statement "running in local mode" as in PR#35112

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35050)

<!-- Reviewable:end -->

JacobTanenbaum committed:
fixed some issues with kubectl set resources

when using kubectl set resources it resets all resource fields that are not being set.

for example
  # kubectl set resources deployments nginx --limits=cpu=100m

followed by

  # kubectl set resources deployments nginx --limits=memory=256Mi

would result in the nginx deployment only limiting memory at 256Mi with the previous
limit placed on the cpu being wiped out. This behavior is corrected so that each invocation
only modifies fields set in that command and changed the testing so that the desired behavior
is checked.

Also a typo:

 you must specify an update to requests or limits or  (in the form of --requests/--limits)

corrected to

 you must specify an update to requests or limits (in the form of --requests/--limits)

changelog:
 - fixed a typo in hack/make-rules/test-cmd.sh "effecting" to "affecting"

janetkuo commented on hack/make-rules/test-cmd.sh:
```diff
@@ -2168,12 +2168,25 @@ __EOF__
+  kubectl set resources deployment nginx-deployment --limits=cpu=200m "${kube_flags[@]}"
+  kube::test::get_object_assert deployment "{{range.items}}{{(index .spec.template.spec.containers 0).resources.limits.cpu}}:{{end}}" "200m:"
+  kube::test::get_object_assert deployment "{{range.items}}{{(index .spec.template.spec.containers 1).resources.limits.cpu}}:{{end}}" "200m:"
+  # Set the deployments memory limits without effecting the cpu limits
```
affecting

janetkuo commented:
Other than typo this LGTM

janetkuo commented:
@k8s-bot ok to test

deads2k commented:
`--local` and `--dry-run` do two different things.  `--local` means, "pull the object from a local file, make the change, show me the output".  `--dry-run` means, "pull the resource *names* from a local file, pull the object from the server, make the change, show me the output (no mutation)"

JacobTanenbaum commented:
@deads2k took out the local vs. dry-run stuff how does it look?

@deads2k @smarterclayton Do we want a --dry-run or a --local flag for this command? (the changes of which would come later)

smarterclayton commented:
All "set" commands should have dry run and local probably.  Users should be able to chain set commands via unix pipes.

JacobTanenbaum committed:
Implemented both the dry run and local flags.

Added test cases to show that both flags are operating as intended.
Removed the print statement "running in local mode" as in PR#35112
